content/posts/Devash_Recommendation_System_Final_Report.md

```diff
@@ -0,0 +1,141 @@
+
+A SQL query was used to extract the id, author_id, name, description, tag_names, views, and star count which were stored as a JSON file (1,80,000 projects approx). We removed the following:
+
+- HTML tags
+- Non-ASCII characters
+- Punctuations.
+- Replaced NULL values and "Untitled" by empty strings.
+
+  > **Note:** We did not remove punctuations like &, \*, +, etc as they have a meaning in logic circuits.
+
+## Feature Extraction
+
+Feature Extraction is mainly done to extract important parts or features from a huge chunk of items, based on the frequency/count of occurrence and thus reducing the dimensions of the dataset to enhance processing.
+Our recommendation system uses [CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) which converts the collection of documents to a matrix of token counts, thus filtering out the important and relevant keywords.
+
+## Topic Modelling using Latent Dirichlet Allocation (LDA)
+
+[LDA](https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html) is used to classify text in a document to a particular topic. It builds a topic per document model and words per topic model, modeled as [Dirichlet distributions](https://www.sciencedirect.com/topics/mathematics/dirichlet-distribution).
+
+- Each document is modeled as a multinomial distribution of topics and each topic is modeled as a multinomial distribution of words.
+- LDA assumes that every chunk of text we feed into it will contain words that are somehow related. Therefore choosing the right corpus of data is crucial.
+- It also assumes documents are produced from a mixture of topics. Those topics then generate words based on their probability distribution.
+
+Simply put, LDA creates a distribution of words and packs them together as a topic (the number of topics is chosen by the user). When a new sentence is fed into the model, based on the words present in the new sentence, we get a distribution of the sentence over all the topics. The new dimension of each vector becomes the total number of topics hence drastically reducing the dimensions with better filtering.
+In our recommendation system, the log-likelihood was the least for 10 topics for the given dataset.
+
+![Topics 5 to 50](/images/DevanshD3-GSoC'21/MT.png)
+**Topics 5 to 50**
+![Zoomed for topics 5 to 20](/images/DevanshD3-GSoC'21/LT.png)
+**Zoomed for topics 5 to 20**
+
+## K-D Trees
+
+After getting the probability distribution of each project we can compare the items and thus enhancing the 2nd layer of the model. Item distance being the best collaborative filtering approach gives us the most similar items. However, this will be a computationally expensive method with time complexity of `O(N^2 log(N))`, and training the model will take days. We also won't be able to use this method for new project recommendations in the online mode because of the high runtime.
```

Comment:
You havent mentioned the two layered model in the beginning

Reply:
Did you mention it?

Reply:
Yes yes I  have incorporated all the changes

Reply:
https://deploy-preview-92--cvblog.netlify.app/posts/devash_recommendation_system_final_report/ 
this is the final report